### PR TITLE
feat: [FFM-7933]: archive instead of delete btn

### PR DESCRIPTION
**Pull request by crystal-fong**
This PR introduces archiving flags instead of deleting a flag.
User will have to type in the flag ID before they are allowed to archive.

Note: The payload we send to the backend will remain the same, currently the archive/soft delete is not in QA yet hence clicking archive will result in a flag being deleted.

The same Archive button will be replaced within the flag detail page at a later time
#### Screenshots

NA

https://github.com/harness/harness-core-ui/assets/95680496/ee54f537-f624-4cc1-829d-4039aac6ddce

#### PR Checklist

<details>
<summary>Please check if your PR fulfils the following requirements</summary>

- [ ] Tests for the changes have been added. Ideally, include a test that fails without this PR but passes with it.
- [ ] Docs have been [added/updated](https://harness.atlassian.net/jira/software/c/projects/DOC/boards/40).
</details>

<details>
<summary>Use the following comments to re-trigger PR Checks</summary>

- Jest: `retrigger jest`
- Prettier: `retrigger prettier`
- Type Check: `retrigger typecheck`
- ESLint: `retrigger eslint`
- Sonar: `retrigger sonar`
- Standards: `retrigger standards`
- Build: `retrigger build`
- Title Check: `retrigger titlecheck`
- Feature Name Check: `trigger featurenamecheck`
- Coverage: `retrigger coverage`
- Rebase: `trigger rebase`
- Cypress Rest: `retrigger cypress-rest`
- Cypress CD: `retrigger cypress-cd`
- Cypress Pipeline: `retrigger cypress-pipeline`
- Cypress CV: `retrigger cypress-cv`
- Fix Prettier: `fix prettier`
</details>

### CODE REVIEW

File: src/modules/75-cf/components/FlagArchiving/useArchiveFlagDialog.tsx [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+          vars={{ flagId: flagData.identifier }}
+          useRichText
+        />
+        <Text>{getString('cf.featureFlags.archiving.confirmFlag')}</Text>
+        <TextInput
+          onPaste={e => e.preventDefault()}
+          errorText={
+            !nameMatch && validationFlagName !== ''
+              ? getString('cf.featureFlags.archiving.mismatchNameError')
+              : undefined
```
Comment (crystal-fong) [inline src/modules/75-cf/components/FlagArchiving/useArchiveFlagDialog.tsx:73]: have to do ternary here otherwise it gives TS warning

File: src/modules/75-cf/components/FlagArchiving/useArchiveFlagDialog.tsx [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    titleText: getString('cf.featureFlags.archiving.archiveFlag'),
+    contentText: (
+      <Layout.Vertical spacing="small">
+        <String
+          stringID="cf.featureFlags.archiving.warningDescription"
+          vars={{ flagId: flagData.identifier }}
+          useRichText
+        />
+        <Text>{getString('cf.featureFlags.archiving.confirmFlag')}</Text>
+        <TextInput
```
Comment (crystal-fong) [inline src/modules/75-cf/components/FlagArchiving/useArchiveFlagDialog.tsx:68]: I decided to not use Formik here as it seemed overkill for just one field

File: src/modules/75-cf/components/FlagArchiving/useArchiveFlagDialog.tsx [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+      handleResponseError(e)
+    }
+  }
+
+  const validateForm = (e: FormEvent<HTMLElement>): void => {
+    if ((e?.target as HTMLInputElement)?.value === flagData?.identifier) {
+      setNameMatch(true)
+    } else {
+      setNameMatch(false)
+    }
```
Comment (Kira-Jennings) [inline src/modules/75-cf/components/FlagArchiving/useArchiveFlagDialog.tsx:54]: Might be able to shorten this to
`setNameMatch((e?.target as HTMLInputElement)?.value === flagData?.identifier)`
If TypeScript doesn't complain about the e.target being possibly undefined

**Review by Kira-Jennings — DISMISSED** [review]
Looks good!

File: src/modules/75-cf/components/FlagArchiving/useArchiveFlagDialog.tsx (hunk 3 above)
Comment (crystal-fong) [reply src/modules/75-cf/components/FlagArchiving/useArchiveFlagDialog.tsx:54]: that is beaut, will update that

Comment (laura-mannus) [reply src/modules/75-cf/components/FlagArchiving/useArchiveFlagDialog.tsx]: chefs kiss 👌

File: src/modules/75-cf/components/FlagArchiving/__tests__/ArchiveFlagButtons.test.tsx [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+      path="/account/:accountId/cf/orgs/:orgIdentifier/projects/:projectIdentifier/feature-flags"
+      pathParams={{ accountId: 'dummy', orgIdentifier: 'dummy', projectIdentifier: 'dummy' }}
+    >
+      <ArchiveFlagButtons nameMatch={false} onArchive={onArchiveMock} onClose={onCloseMock} {...props} />
+    </TestWrapper>
+  )
+}
+
+describe('ArchiveFlagButtons', () => {
+  test('it should disable the Archive button if the user types in a mismatching flag Id', async () => {
```
Comment (laura-mannus) [inline src/modules/75-cf/components/FlagArchiving/__tests__/ArchiveFlagButtons.test.tsx:22]: I am confusion with capitalisations...should this be?

File: src/modules/75-cf/components/FlagArchiving/__tests__/useArchiveFlagDialog.test.tsx [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    const incorrectFlagId = 'foobar'
+    renderComponent()
+
+    userEvent.type(screen.getByRole('textbox'), incorrectFlagId)
+
+    expect(screen.getByRole('textbox')).toHaveValue(incorrectFlagId)
+    expect(screen.getByText('cf.featureFlags.archiving.mismatchNameError')).toBeInTheDocument()
+    expect(screen.getByRole('button', { name: 'archive' })).toBeDisabled()
+
+    userEvent.clear(screen.getByRole('textbox'))
```
Comment (laura-mannus) [inline src/modules/75-cf/components/FlagArchiving/__tests__/useArchiveFlagDialog.test.tsx:55]: NBD: Because you are using `screen.getByRole('textbox`) here a lot for readability you could extract it to a constant. But totally preference.

File: src/modules/75-cf/components/FlagArchiving/useArchiveFlagDialog.tsx [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+  openDialog: () => void
+}
+
+const useArchiveFlagDialog = ({
+  flagData,
+  deleteFeatureFlag,
+  queryParams,
+  refetchFlags
+}: ArchiveDialogProps): UseArchiveFlagDialogReturn => {
+  const [isANameMatch, setIsANameMatch] = useState<boolean>(false)
```
Comment (laura-mannus) [inline src/modules/75-cf/components/FlagArchiving/useArchiveFlagDialog.tsx:31]:
```suggestion
  const [isIdMatch, setIsIdMatch] = useState<boolean>(false)
```
should this be `isANameMatch ` as we are actually asking the user to type ID?

File: src/modules/75-cf/components/FlagArchiving/useArchiveFlagDialog.tsx [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+  const [validationFlagName, setValidationFlagName] = useState<string>('')
```
Comment (laura-mannus) [inline src/modules/75-cf/components/FlagArchiving/useArchiveFlagDialog.tsx:32]:
```suggestion
  const [validationFlagId, setValidationFlagId] = useState<string>('')
```
same here ?

File: src/modules/75-cf/components/FlagArchiving/useArchiveFlagDialog.tsx [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+          placeholder={flagData?.identifier}
+          value={validationFlagName}
+          onChange={e => {
+            setValidationFlagName((e?.target as HTMLInputElement)?.value)
+            setIsANameMatch((e?.target as HTMLInputElement)?.value === flagData?.identifier)
+          }}
+        />
+      </Layout.Vertical>
+    ),
+    customButtons: (
```
Comment (laura-mannus) [inline src/modules/75-cf/components/FlagArchiving/useArchiveFlagDialog.tsx:78]: I love this prop

File: src/modules/75-cf/strings/strings.en.yaml [hunk 9]
```diff
@@ -165,14 +165,13 @@ featureFlags:
   archiving:
     archiveFlag: Archive Flag
     archiveSuccess: Flag successfully archived
-    cannotArchive: '{{ flagName }} cannot be archived because it is still being used by the following flags as a prerequisite'
-    confirmFlag: Type the flag name to confirm
-    mismatchNameError: The flag name does not match
+    cannotArchive: '{{ flagId }} cannot be archived because it is still being used by the following flags as a prerequisite'
+    confirmFlag: Type the flag Id to confirm
+    mismatchNameError: The flag Id does not match
```
Comment (laura-mannus) [inline src/modules/75-cf/strings/strings.en.yaml:170]:
```suggestion
    mismatchIdError: The flag Id does not match
```

File: src/modules/75-cf/components/FlagArchiving/ArchiveFlagButtons.tsx [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+  onArchive: () => void
+  onClose: () => void
+}
+
+const ArchiveFlagButtons: FC<ArchiveFlagButtonsProps> = ({ onArchive, idMatch, onClose }) => {
+  const { getString } = useStrings()
+
+  return (
+    <Layout.Horizontal spacing="small" flex={{ alignItems: 'flex-start' }}>
+      <Button disabled={!idMatch} text={getString('archive')} intent={Intent.DANGER} onClick={onArchive} />
```
Comment (knagurski) [inline src/modules/75-cf/components/FlagArchiving/ArchiveFlagButtons.tsx:17]: Remember to use the primary variation on this.

File: src/modules/75-cf/components/FlagArchiving/CannotArchiveWarning.tsx [hunk 11]
```diff
@@ -16,19 +16,19 @@ import type { Prerequisite } from 'services/cf'
 import { ItemContainer } from '../ItemContainer/ItemContainer'
 
 export interface CannotArchiveWarningProps {
-  flagName: string
+  flagId: string
```
Comment (knagurski) [inline src/modules/75-cf/components/FlagArchiving/CannotArchiveWarning.tsx:19]: _SUPER_ not a blocker, but for your own sanity, try to stick with `thingIdentifier` instead of `thingId`. The only legit place we use `Id` is when getting the `accountId` from the route params. It'll confuse you down the road, so just a good habit to get into. Again, 100% not a blocker.

File: src/modules/75-cf/components/FlagArchiving/__tests__/ArchiveFlagButtons.test.tsx (hunk 4 above)
Comment (knagurski) [reply src/modules/75-cf/components/FlagArchiving/__tests__/ArchiveFlagButtons.test.tsx:22]: In test names don't worry about capitalisation too much. They're for our use, so if having a word capitalised emphasises it better, fire away. We only need to worry in customer-facing text.

File: src/modules/75-cf/components/FlagArchiving/__tests__/ArchiveFlagButtons.test.tsx [hunk 12]
```diff
@@ -0,0 +1,49 @@
+import React from 'react'
+import { render, RenderResult, screen, waitFor } from '@testing-library/react'
+import userEvent from '@testing-library/user-event'
+import { TestWrapper } from '@common/utils/testUtils'
+import ArchiveFlagButtons, { ArchiveFlagButtonsProps } from '../ArchiveFlagButtons'
+
+const onArchiveMock = jest.fn()
+const onCloseMock = jest.fn()
```
Comment (knagurski) [inline src/modules/75-cf/components/FlagArchiving/__tests__/ArchiveFlagButtons.test.tsx:8]: I would avoid these out-of-scope mocks that are asserted on. As-is there's a bug below that would be apparent if you swapped the order of the first and second tests. You _could_ fix it with a `jest.clearAllMocks()` in a `beforeEach`, but a more fool proof way is to pass in things you care about in the test you care about it in.

So for your `renderComponent` you'd replace `onArchive={onArchiveMock}` and `onClose={onCloseMock}` with `onArchive={jest.fn()}` and `onClose={jest.fn()}` and then in your test, just create it there and pass it into the `renderComponent` call. It'll then be impossible to leak state and have tests that are either flaky or break when run in a different order.

This is what happens when you swap the 1st and 2nd tests.
Screenshot 2023-06-09 at 12 39 13

File: src/modules/75-cf/components/FlagArchiving/__tests__/ArchiveFlagButtons.test.tsx [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+
+  test('it should call onArchive callback if user types in matching flag Id', async () => {
+    renderComponent({ idMatch: true })
+
+    expect(screen.getByRole('button', { name: 'archive' })).toBeEnabled()
+
+    userEvent.click(screen.getByRole('button', { name: 'archive' }))
+
+    await waitFor(() => expect(onArchiveMock).toHaveBeenCalled())
+  })
```
Comment (knagurski) [inline src/modules/75-cf/components/FlagArchiving/__tests__/ArchiveFlagButtons.test.tsx:40]: As mentioned above. Try to avoid out-of-scope things you assert on. If you were to kill off the `onArchiveMock` above the tests, you can rewrite this test pretty easily.
```suggestion
  test('it should call onArchive callback if user types in matching flag Id', async () => {
    const onArchiveMock = jest.fn()
    renderComponent({ idMatch: true, onArchive: onArchiveMock })

    expect(screen.getByRole('button', { name: 'archive' })).toBeEnabled()

    userEvent.click(screen.getByRole('button', { name: 'archive' }))

    await waitFor(() => expect(onArchiveMock).toHaveBeenCalled())
  })
```

File: src/modules/75-cf/strings/strings.en.yaml [hunk 14]
```diff
@@ -165,14 +165,13 @@ featureFlags:
   archiving:
     archiveFlag: Archive Flag
     archiveSuccess: Flag successfully archived
-    cannotArchive: '{{ flagName }} cannot be archived because it is still being used by the following flags as a prerequisite'
-    confirmFlag: Type the flag name to confirm
-    mismatchNameError: The flag name does not match
+    cannotArchive: '{{ flagId }} cannot be archived because it is still being used by the following flags as a prerequisite'
+    confirmFlag: Type the flag Id to confirm
+    mismatchIdError: The flag Id does not match
```
Comment (knagurski) [inline src/modules/75-cf/strings/strings.en.yaml:170]: Try to stick to `identifier` to match the wording of the rest of Harness.
```suggestion
    confirmFlag: Type the flag identifier to confirm
    mismatchIdError: The flag identifier does not match
```

**Review by Kira-Jennings — APPROVED** [review]

**Review by laura-mannus — APPROVED** [review]
Really fab job
